### DISCUSSION TITLE
Is Internationalization Date package a replacement for Moment.js & Date-Fns?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [adobe/react-spectrum#2775] [Q&A] [answered]
Same.

**Comment by devongovett** [comment, marked as answer]
Yes. It is similar to other date libraries, but focused on internationalization (i.e. working with different calendar systems, time zones, etc.). These use cases are not typically covered by other libraries. If you're curious, I'm currently working on the documentation [here](https://reactspectrum.blob.core.windows.net/reactspectrum/04add13580b7473926413f3c2457bd2119cd272a/docs/internationalized/date/index.html).

**Reply by deadcoder0904** [reply]
oh yes, i've already seen the docs & love how intuitive everything is.

**Reply by barbinbrad** [reply]
I'd like our company to move to `@internationalized/date` -- but one of the big push-backs has been on formatting. For example converting a `ZonedDateTime` to a string like `3 minutes ago`. `date-fns` seems to excel here. But from my perspective, this is the trivial part of a date time library.

Two questions:

1. Is [this type of formatting](https://github.com/date-fns/date-fns/blob/main/src/formatDistance/index.ts) something you'd be interested in adding to the `@internationalized/date` package?
2. How would you go about implementing the different translations. [For example](https://github.com/date-fns/date-fns/blob/main/src/locale/es/_lib/formatDistance/index.ts).

**Reply by devongovett** [reply]
For formatting, you should use the native browser date formatting APIs. Using these means there is no hit to bundle size, and you get full internationalization capabilities.

To format dates and times, use [Intl.DateTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat).

To format relative times, use [Intl.RelativeTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat).

You can format one of the date objects from `@internationalized/date` by first converting it to a native JS Date object. For example, with a `ZonedDateTime`, there is a [toDate()](https://react-spectrum.adobe.com/internationalized/date/ZonedDateTime.html#to-a-native-date-object) method.

**Reply by barbinbrad** [reply]
Awesome! Thanks for sharing that. So if I'm reading this correctly, the `format` method of `Intl.RelativeTimeFormat` still needs a transform step to go from date to an argument like this `(-3, 'days')`? Does something like this exist already in `@internationalized/date`?


```js
const formatter = new Intl.RelativeTimeFormat(undefined, {
  numeric: 'auto'
})

const DIVISIONS = [
  { amount: 60, name: 'seconds' },
  { amount: 60, name: 'minutes' },
  { amount: 24, name: 'hours' },
  { amount: 7, name: 'days' },
  { amount: 4.34524, name: 'weeks' },
  { amount: 12, name: 'months' },
  { amount: Number.POSITIVE_INFINITY, name: 'years' }
]

function formatTimeAgo(date) {
  let duration = (date - new Date()) / 1000

  for (let i = 0; i <= DIVISIONS.length; i++) {
    const division = DIVISIONS[i]
    if (Math.abs(duration) < division.amount) {
      return formatter.format(Math.round(duration), division.name)
    }
    duration /= division.amount
  }
}
```

**Reply by devongovett** [reply]
At the moment, there is nothing built in for that. Ideally, this would use the calendar system to do this rather than hard coding the number of months/weeks etc. In the meantime, you could easily use the above function by converting to JS date objects.

**Reply by rostero1** [reply]
Do you have an example of what this would look like with `@internationalized/date` or has there been an update to add relative time (or to get the difference between two dates)?
